src/extract_gpuinfo_ascend.c

```diff
@@ -0,0 +1,251 @@
+  return last_dcmi_return_status == DCMI_SUCCESS;
+}
+
+static void gpuinfo_ascend_shutdown(void) {
+  local_error_string = "";
+
+  struct gpu_info_ascend *allocated, *tmp;
+  list_for_each_entry_safe(allocated, tmp, &allocations, allocate_list) {
+    list_del(&allocated->allocate_list);
+    free(allocated);
+  }
+}
+
+static const char *gpuinfo_ascend_last_error_string(void) {
+  return local_error_string;
+}
+
+static bool gpuinfo_ascend_get_device_handles(struct list_head *devices, unsigned *count) {
+  int num_cards;
+  int card_list[MAX_CARD_NUM] = {0};
+  last_dcmi_return_status = dcmi_get_card_list(&num_cards, card_list, MAX_DEVICE_NUM);
+  if (last_dcmi_return_status != DCMI_SUCCESS) {
+    local_error_string = "Failed to get card num";
+    return false;
+  } else if (num_cards == 0) {
+    local_error_string = "Not found NPU(s)";
+    return false;
+  }
+
+  int num_devices = 0;
+  int card_device_list[num_cards];
+  for (int i = 0; i < num_cards; ++i) {
+    int num_card_devices;
+    last_dcmi_return_status = dcmi_get_device_num_in_card(card_list[i], &num_card_devices);
+    if (last_dcmi_return_status != DCMI_SUCCESS) {
+      local_error_string = "Failed to get device num of card";
+      return false;
+    }
+    num_devices += num_card_devices;
+    card_device_list[i] = num_card_devices;
+  }
+
+  struct gpu_info_ascend *gpu_infos = calloc(num_devices, sizeof(*gpu_infos));
+  if (!gpu_infos) {
+    local_error_string = strerror(errno);
+    return false;
+  }
+
+  // todo: for free gpu_infos when shutting down, rewrite to direct free?
+  list_add(&gpu_infos[0].allocate_list, &allocations);
+
+  *count = 0;
+  for (int i = 0; i < num_cards; ++i) {
+    for (int j = 0; j < card_device_list[i]; ++j) {
+      gpu_infos[*count].base.vendor = &gpu_vendor_ascend;
+      _encode_card_device_id_to_pdev(gpu_infos[*count].base.pdev, i, j);
+      list_add_tail(&gpu_infos[*count].base.list, devices);
+      *count += 1;
+    }
+  }
+
+  return true;
+}
+
+static void _encode_card_device_id_to_pdev(char *pdev, int card_id, int device_id) {
+  sprintf(pdev, "%d-%d", (short)card_id, (short)device_id);
+}
+
+static void _decode_card_device_id_from_pdev(const char *pdev, int *card_id, int *device_id) {
+  sscanf(pdev, "%d-%d", card_id, device_id);
+}
+
+static void gpuinfo_ascend_populate_static_info(struct gpu_info *_gpu_info) {
+  struct gpu_info_ascend *gpu_info = container_of(_gpu_info, struct gpu_info_ascend, base);
+  struct gpuinfo_static_info *static_info = &gpu_info->base.static_info;
+  static_info->integrated_graphics = false;
+  RESET_ALL(static_info->valid);
+
+  int card_id, device_id;
+  _decode_card_device_id_from_pdev(_gpu_info->pdev, &card_id, &device_id);
+
+  struct dcmi_chip_info *chip_info = malloc(sizeof(struct dcmi_chip_info));
+  last_dcmi_return_status = dcmi_get_device_chip_info(card_id, device_id, chip_info);
+  if (last_dcmi_return_status == DCMI_SUCCESS) {
+    // assume Ascend only use ASCII code for chip name
```

Comment:
I think that libncurses will display non-ascii correctly if the terminal supports it.

Reply:
The next line `strcpy(static_info->device_name, (char *) chip_info->chip_name);`, in order to clean the warning I  add a cast from unsigned char to char, but if the Ascend chip name is in UTF-8 charset, there may cause incorrect result (hardly happen)